### PR TITLE
Stats v4 UI: show more granular labels on chart x-axis

### LINKED ISSUE #1178: Stats UI: update x-axis labels
For stats v4 UI, we show x-axis labels more than just the start and end dates. Different number of labels is shown for each tab (today/this week/this month/this year), and the month name is only shown for the first day of a unique month.

For "today" tab where we display the hours, the text could be longer in some languages like Chinese and Japanese. Let's try rotating the labels if the width of all labels is greater than the chart width. If it's not working after like 1-hour time box, then we just show a fewer number of labels.

### PR DESCRIPTION
**Pull request by jaclync**
Fixes #1178 

- [x] ⚠️ update PR base to `develop` before merging ⚠️ 

## Changes
- Created `StoreStatsV4ChartAxisHelper` that returns the number of labels to show on x-axis, and the text of each label
  - Added a test case on the more complicated scenario when we only show the month of the date for the first date of the month (e.g. for dates like `[July 31, August 1, August 2]` we show `["July 31", "August 1", "2"]` respectively)
- Added `orderStatsIntervalLabels: [String]` in `StoreStatsV4PeriodViewController` that gets updated whenever `orderStatsIntervals` is set with the label text from `StoreStatsV4ChartAxisHelper`
- Note: I reduced the number of labels in "today" tab from 6 to 5 because the labels overlap in certain languages like zh-Hant. I tried turning on `wordWrapEnabled` for x-axis but I couldn't get the second line of text to show - if you have any experience with word wrap in `Charts`, I'd love to show the label text in multiple lines instead! I also time-boxed attempting to rotate the text when the total label width exceeds the chart width, but couldn't find a good way to get the label/chart width.

## Testing

Prerequisite: the store has WC admin plugin activated
- Go to "My store"
- Check the x-axis of the stats chart in each time range --> the labels should look the same as in design

## Example screenshots

locale | today | this week | this month | this year
--- | --- | --- | --- | ---
en | Simulator Screen Shot - iPhone 5s - 2019-08-21 at 10 16 00 | Simulator Screen Shot - iPhone 5s - 2019-08-21 at 10 16 03 | Simulator Screen Shot - iPhone 5s - 2019-08-21 at 10 16 07 | Simulator Screen Shot - iPhone 5s - 2019-08-21 at 10 16 10
zh-Hant | Simulator Screen Shot - iPhone 5s - 2019-08-21 at 10 16 43 | Simulator Screen Shot - iPhone 5s - 2019-08-21 at 10 16 46 | Simulator Screen Shot - iPhone 5s - 2019-08-21 at 10 16 49 | Simulator Screen Shot - iPhone 5s - 2019-08-21 at 10 16 50

Update release notes:

- [x] If there are user facing changes, I have added an item to `RELEASE-NOTES.txt`.

### CODE REVIEW

File: WooCommerce/WooCommerceTests/ViewRelated/Dashboard/StoreStatsV4ChartAxisHelperTests.swift [hunk 1]
```diff
@@ -0,0 +1,38 @@
...
+class StoreStatsV4ChartAxisHelperTests: XCTestCase {
+    private let helper = StoreStatsV4ChartAxisHelper()
+
+    func testDatesAcrossTwoMonths() {
+        // GMT: Saturday, June 1, 2019 12:29:29 AM
+        let dateInJune = Date(timeIntervalSince1970: 1559348969)
+        // GMT: Monday, June 10, 2019 12:29:29 AM
+        let secondDateInJune = Date(timeIntervalSince1970: 1560126569)
+        // GMT: Thursday, August 1, 2019 12:29:29 AM
+        let dateInAugust = Date(timeIntervalSince1970: 1561940969)
```
Comment (mindgraffiti) [inline WooCommerce/WooCommerceTests/ViewRelated/Dashboard/StoreStatsV4ChartAxisHelperTests.swift]: Did you mean to test August 1st or July 1st? 

>Monday, July 1, 2019 12:29:29 AM

Both are consistent July 1st dates based on Epoch time, so the unit tests pass.

**Comment by mindgraffiti** [comment]
x-axis labels - this one also looks relevant https://github.com/danielgindi/Charts/issues/3933

File: WooCommerce/Classes/ViewRelated/Dashboard/Stats v4/StoreStatsV4ChartAxisHelper.swift [hunk 2]
```diff
@@ -0,0 +1,51 @@
+import Yosemite
+
+/// A helper for the labels for store stats v4 charts UI.
+class StoreStatsV4ChartAxisHelper {
```
Comment (ctarda) [inline WooCommerce/Classes/ViewRelated/Dashboard/Stats v4/StoreStatsV4ChartAxisHelper.swift]: I'm just going to leave my usual comment... 😛  Do we really want to be able to subclass this? Should it be final?

File: WooCommerce/Classes/ViewRelated/Dashboard/Stats v4/StoreStatsV4ChartAxisHelper.swift [hunk 3]
```diff
@@ -0,0 +1,51 @@
...
+        }
+
+        var dateLabelText = [String]()
+
+        var calendar = Calendar.current
+        calendar.timeZone = siteTimezone
+        var latestMonth: Int? = nil
+        let dayOfMonthDateFormatter = DateFormatter.Charts.chartAxisDayOfMonthFormatter
+        dayOfMonthDateFormatter.timeZone = siteTimezone
+        for date in dates {
```
Comment (ctarda) [inline WooCommerce/Classes/ViewRelated/Dashboard/Stats v4/StoreStatsV4ChartAxisHelper.swift]: Just out of curiosity, and for kicks and giggles, I tried to write this using reduce, and it was way more compact, but also way more unreadable (at least to me)

File: WooCommerce/Classes/ViewRelated/Dashboard/Stats v4/StoreStatsV4ChartAxisHelper.swift (hunk 2 above)
Comment (jaclync) [reply WooCommerce/Classes/ViewRelated/Dashboard/Stats v4/StoreStatsV4ChartAxisHelper.swift]: my bad! I certainly forgot the new habit to mark a class as final by default 😬will update!

File: WooCommerce/WooCommerceTests/ViewRelated/Dashboard/StoreStatsV4ChartAxisHelperTests.swift (hunk 1 above)
Comment (jaclync) [reply WooCommerce/WooCommerceTests/ViewRelated/Dashboard/StoreStatsV4ChartAxisHelperTests.swift]: ohh great catch, I'll update the timestamp!

File: WooCommerce/Classes/ViewRelated/Dashboard/Stats v4/StoreStatsV4ChartAxisHelper.swift (hunk 3 above)
Comment (jaclync) [reply WooCommerce/Classes/ViewRelated/Dashboard/Stats v4/StoreStatsV4ChartAxisHelper.swift]: I like reduce! went for the most straightforward for loop first, lemme rewrite with reduce

**Comment by jaclync** [comment]
@mindgraffiti thanks for linking the issues! I created this issue for potential improvements in the future https://github.com/woocommerce/woocommerce-ios/issues/1207

@ctarda @mindgraffiti I updated the PR based on your suggestions, and for @kyleaparker to take a look!

**Comment by mindgraffiti** [comment]
From @kyleaparker 

>just wondering what text size is being used as the default?

@jaclync I didn't catch that earlier. The `chartLabelFont` weight should not be set to `ultraLight`, it should be `regular`. 

```
let chartLabelFont                  = UIFont.font(forStyle: .caption2, weight: .regular)
```

:shipit: after ☝️

**Comment by jaclync** [comment]
@kyleaparker @mindgraffiti updated the chart label font from `ultra-light` to `regular` `caption2` (screenshots in the PR description also updated), merging now
@mindgraffiti thanks for checking the font!

**Pull request merged**